csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Ipc
+{
+    internal class ArrowStreamReaderImplementation : ArrowReaderImplementation
+    {
+        public Stream BaseStream { get; }
+        protected ArrayPool<byte> Buffers { get; }
+
+        public ArrowStreamReaderImplementation(Stream stream)
+        {
+            BaseStream = stream;
+            Buffers = ArrayPool<byte>.Create();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                BaseStream.Dispose();
+            }
+        }
+
+        public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken)
+        {
+            // TODO: Loop until a record batch is read.
+            cancellationToken.ThrowIfCancellationRequested();
+            return await ReadRecordBatchAsync(cancellationToken);
+        }
+
+        public override RecordBatch ReadNextRecordBatch()
+        {
+            throw new NotImplementedException();
+        }
+
+        protected async Task<RecordBatch> ReadRecordBatchAsync(CancellationToken cancellationToken = default)
+        {
+            await ReadSchemaAsync();
+
+            var bytesRead = 0;
+
+            byte[] lengthBuffer = null;
+            byte[] messageBuff = null;
+            byte[] bodyBuff = null;
+
+            try
+            {
+                // Get Length of record batch for message header.
+
+                lengthBuffer = Buffers.Rent(4);
+                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken);
```

Comment:
As noted earlier, especially since this is dealing with an arbitrary stream, it'd be better to use the Memory-based overloads for all of these ReadAsyncs.

Reply:
Same about [ARROW-4503: ArrowStreamReader allocates and copies data excessively](https://issues.apache.org/jira/browse/ARROW-4503)